TCP assertion failure in cm

http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/9273/pipeline/232

```
[2021-01-19T22:27:28.738Z] [ RUN      ] all/test_ucp_sockaddr_destroy_ep_on_err.onesided_client_sforce/1 <all/tag,mt>
[2021-01-19T22:27:28.738Z] [     INFO ] Testing 10.209.44.19:0
[2021-01-19T22:27:28.738Z] [     INFO ] server listening on 10.209.44.19:41369
[2021-01-19T22:27:28.738Z] [     INFO ] < UCT EP 0xffff300011a0 isn't associated with UCP EP and was not scheduled to be discarded on UCP Worker 0x10290f0 >
[2021-01-19T22:27:28.738Z] [hpc-arm-cavium-jenkins:23377:0:23377]      tcp_cm.c:83   Assertion `(old_conn_state == UCT_TCP_EP_CONN_STATE_RECV_MAGIC_NUMBER) || ((old_conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) && (ep->conn_retries == 0) && (ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP))' failed
[2021-01-19T22:27:28.738Z] 
[2021-01-19T22:27:28.738Z] /scrap/jenkins/workspace/ucx-8/contrib/../src/uct/tcp/tcp_cm.c: [ uct_tcp_cm_change_conn_state() ]
[2021-01-19T22:27:28.738Z]       ...
[2021-01-19T22:27:28.738Z]        77         break;
[2021-01-19T22:27:28.738Z]        78     case UCT_TCP_EP_CONN_STATE_ACCEPTING:
[2021-01-19T22:27:28.738Z]        79         ucs_assert((old_conn_state == UCT_TCP_EP_CONN_STATE_RECV_MAGIC_NUMBER) ||
[2021-01-19T22:27:28.738Z] ==>    80                    ((old_conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) &&
[2021-01-19T22:27:28.738Z]        81                     (ep->conn_retries == 0) &&
[2021-01-19T22:27:28.738Z]        82                     (ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP)));
[2021-01-19T22:27:28.738Z]        83         uct_tcp_iface_outstanding_inc(iface);
[2021-01-19T22:27:28.738Z] 
[2021-01-19T22:27:30.652Z] ==== backtrace (tid:  23377) ====
[2021-01-19T22:27:30.652Z]  0 0x0000000000056b90 ucs_debug_print_backtrace()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/debug/debug.c:656
[2021-01-19T22:27:30.652Z]  1 0x0000000000028b44 uct_tcp_cm_change_conn_state()  /scrap/jenkins/workspace/ucx-8/contrib/../src/uct/tcp/tcp_cm.c:80
[2021-01-19T22:27:30.652Z]  2 0x0000000000020be4 uct_tcp_ep_connect_to_ep()  /scrap/jenkins/workspace/ucx-8/contrib/../src/uct/tcp/tcp_ep.c:725
[2021-01-19T22:27:30.652Z]  3 0x000000000001517c uct_ep_connect_to_ep()  /scrap/jenkins/workspace/ucx-8/contrib/../src/uct/base/uct_iface.c:489
[2021-01-19T22:27:30.652Z]  4 0x000000000001517c uct_ep_connect_to_ep()  /scrap/jenkins/workspace/ucx-8/contrib/../src/uct/base/uct_iface.c:489
[2021-01-19T22:27:30.652Z]  5 0x0000000000093cbc ucp_wireup_connect_local()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucp/wireup/wireup.c:316
[2021-01-19T22:27:30.652Z]  6 0x0000000000098cf4 ucp_cm_client_connect_progress()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucp/wireup/wireup_cm.c:468
[2021-01-19T22:27:30.652Z]  7 0x000000000004c554 ucs_callbackq_slow_proxy()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/datastruct/callbackq.c:400
[2021-01-19T22:27:30.652Z]  8 0x000000000004c554 ucs_callbackq_enter()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/datastruct/callbackq.c:62
[2021-01-19T22:27:30.652Z]  9 0x000000000004c554 ucs_callbackq_slow_proxy()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/datastruct/callbackq.c:402
[2021-01-19T22:27:30.652Z] 10 0x000000000003ac40 ucs_callbackq_dispatch()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucs/datastruct/callbackq.h:211
[2021-01-19T22:27:30.653Z] 11 0x000000000003ac40 uct_worker_progress()  /scrap/jenkins/workspace/ucx-8/contrib/../src/uct/api/uct.h:2436
[2021-01-19T22:27:30.653Z] 12 0x000000000003ac40 ucp_worker_progress()  /scrap/jenkins/workspace/ucx-8/contrib/../src/ucp/core/ucp_worker.c:2407
[2021-01-19T22:27:30.653Z] 13 0x0000000000873e18 ucp_test_base::entity::progress()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/ucp/ucp_test.cc:870
[2021-01-19T22:27:30.653Z] 14 0x0000000000873eac ucp_test::progress()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/ucp/ucp_test.cc:155
[2021-01-19T22:27:30.653Z] 15 0x0000000000857c58 test_ucp_sockaddr::one_sided_disconnect()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:497
[2021-01-19T22:27:30.653Z] 16 0x0000000000847214 test_ucp_sockaddr_destroy_ep_on_err_onesided_client_sforce_Test::test_body()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:913
[2021-01-19T22:27:30.653Z] 17 0x000000000059a43c ucs::test_base::run()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/test.cc:344
[2021-01-19T22:27:30.653Z] 18 0x000000000059a43c ucs::test_base::TestBodyProxy()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/test.cc:370
[2021-01-19T22:27:30.653Z] 19 0x000000000057cb2c HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:3562
[2021-01-19T22:27:30.653Z] 20 0x000000000057cb2c testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:3598
[2021-01-19T22:27:30.653Z] 21 0x00000000005712e4 testing::Test::Run()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:3635
[2021-01-19T22:27:30.653Z] 22 0x00000000005713b4 testing::TestInfo::Run()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:3812
[2021-01-19T22:27:30.653Z] 23 0x0000000000571524 testing::TestCase::Run()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:3930
[2021-01-19T22:27:30.653Z] 24 0x0000000000575934 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:5808
[2021-01-19T22:27:30.653Z] 25 0x0000000000575c50 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:5725
[2021-01-19T22:27:30.653Z] 26 0x0000000000575c50 HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:3618
[2021-01-19T22:27:30.653Z] 27 0x0000000000575c50 testing::UnitTest::Run()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest-all.cc:5422
[2021-01-19T22:27:30.653Z] 28 0x000000000051da78 RUN_ALL_TESTS()  /scrap/jenkins/workspace/ucx-8/contrib/../test/gtest/common/gtest.h:20059
[2021-01-19T22:27:30.653Z] 29 0x00000000000215d4 __libc_start_main()  :0
[2021-01-19T22:27:30.653Z] 30 0x000000000055bc1c _start()  :0
[2021-01-19T22:27:30.653Z] =================================
[2021-01-19T22:27:30.653Z] Sending notification to lgenkin@nvidia.com
[2021-01-19T22:27:38.783Z] [hpc-arm-cavium-jenkins:23377:0:23377] Process frozen...
[2021-01-20T00:52:36.326Z] make: *** [test] Terminated
script returned exit code 124
```

--- a/src/uct/tcp/tcp.h
+++ b/src/uct/tcp/tcp.h
@@ -164,11 +164,11 @@ typedef enum uct_tcp_cm_conn_event {
     /* Connection request from a EP that has TX capability to a EP that
      * has to be able to receive AM data (i.e. has to have RX capability). */
     UCT_TCP_CM_CONN_REQ               = UCS_BIT(0),
-    /* Connection acknowledgment from a EP that accepts a conenction from
+    /* Connection acknowledgment from a EP that accepts a connection from
      * initiator of a connection request. */
     UCT_TCP_CM_CONN_ACK               = UCS_BIT(1),
     /* Connection acknowledgment + Connection request. The mesasge is sent
-     * from a EP that accepts remote conenction when it was in
+     * from a EP that accepts remote connection when it was in
      * `UCT_TCP_EP_CONN_STATE_CONNECTING` state (i.e. original
      * `UCT_TCP_CM_CONN_REQ` wasn't sent yet) and want to have RX capability
      * on a peer's EP in order to send AM data. */

--- a/src/uct/tcp/tcp_ep.c
+++ b/src/uct/tcp/tcp_ep.c
@@ -150,14 +150,24 @@ static void uct_tcp_ep_cleanup(uct_tcp_ep_t *ep)
     ucs_close_fd(&ep->stale_fd);
 }
 
+static void uct_tcp_ep_ptr_map_verify(uct_tcp_ep_t *ep, int on_ptr_map)
+{
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
+    if (on_ptr_map) {
+        ucs_assert(ep->flags & UCT_TCP_EP_FLAG_ON_PTR_MAP);
+    } else {
+        ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_ON_PTR_MAP));
+    }
+    ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_ON_MATCH_CTX));
+}
+
 static void uct_tcp_ep_ptr_map_add(uct_tcp_ep_t *ep)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
     ucs_status_t status;
 
-    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
-    ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_ON_MATCH_CTX));
+    uct_tcp_ep_ptr_map_verify(ep, 0);
 
     status = ucs_ptr_map_put(&iface->ep_ptr_map, ep, 1,
                              &ep->cm_id.ptr_map_key);
@@ -168,9 +178,7 @@ static void uct_tcp_ep_ptr_map_add(uct_tcp_ep_t *ep)
 
 static void uct_tcp_ep_ptr_map_removed(uct_tcp_ep_t *ep)
 {
-    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
-    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_ON_PTR_MAP);
-    ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_ON_MATCH_CTX));
+    uct_tcp_ep_ptr_map_verify(ep, 1);
     ep->flags &= ~UCT_TCP_EP_FLAG_ON_PTR_MAP;
 }
 
@@ -183,6 +191,23 @@ static void uct_tcp_ep_ptr_map_del(uct_tcp_ep_t *ep)
     status = ucs_ptr_map_del(&iface->ep_ptr_map, ep->cm_id.ptr_map_key);
     ucs_assert_always(status == UCS_OK);
     uct_tcp_ep_ptr_map_removed(ep);
+}
+
+static uct_tcp_ep_t *
+uct_tcp_ep_ptr_map_get(uct_tcp_iface_t *iface, ucs_ptr_map_key_t ptr_map_key)
+{
+    ucs_status_t status;
+    uct_tcp_ep_t *ep;
+    void *ptr;
+
+    status = ucs_ptr_map_get(&iface->ep_ptr_map, ptr_map_key, 0, &ptr);
+    if (ucs_likely(status == UCS_OK)) {
+        ep = ptr;
+        uct_tcp_ep_ptr_map_verify(ep, 1);
+        return ep;
+    }
+
+    return NULL;
 }
 
 uct_tcp_ep_t *uct_tcp_ep_ptr_map_retrieve(uct_tcp_iface_t *iface,
@@ -590,7 +615,7 @@ static ucs_status_t uct_tcp_ep_connect(uct_tcp_ep_t *ep)
         }
     } else {
         /* EP that connects to self or EP created using CONNECT_TO_EP mustn't
-         * go here and always create socket and conenct to a peer */
+         * go here and always create socket and connect to a peer */
         ucs_assert(!uct_tcp_ep_is_self(ep) &&
                    !(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP));
         ucs_assert((peer_ep != NULL) && (peer_ep->fd != -1) &&
@@ -709,19 +734,33 @@ ucs_status_t uct_tcp_ep_connect_to_ep(uct_ep_h tl_ep,
         return UCS_OK;
     }
 
+    if (uct_tcp_ep_ptr_map_get(iface, ep->cm_id.ptr_map_key) == NULL) {
+        /* If the EP doesn't exist anymore in the local EP PTR MAP, it means the
+         * EP was already connected to a peer's EP and removed from the EP PTR
+         * MAP as a part of CONN_REQ handling (only not connected yet EPs are
+         * contained in a PTR MAP) and then disconnected before a user called
+         * uct_ep_connect_to_ep() for the EP.
+         * So, just report to a caller that the connection was already reset */
+        ucs_assert(uct_tcp_cm_ep_accept_conn(ep));
+        ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED);
+        ucs_assert(ep->conn_retries > 0);
+        return UCS_ERR_CONNECTION_RESET;
+    }
+
     uct_tcp_ep_set_dest_addr(dev_addr, (uct_iface_addr_t*)&addr->iface_addr,
                              &ep->peer_addr);
 
     if (!uct_tcp_cm_ep_accept_conn(ep)) {
         ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED);
-        /* EP that are created as CONNECT_TO_EP has to be full-duplex, set RX
+        /* EPs which are created as CONNECT_TO_EP has to be full-duplex, set RX
          * capability as well as TX (that's set in uct_tcp_ep_connect()) */
         uct_tcp_ep_add_ctx_cap(ep, UCT_TCP_EP_FLAG_CTX_TYPE_RX);
 
         uct_tcp_ep_ptr_map_del(ep);
 
-        /* Use remote peer connection sequence number value, since the EP has to
-         * send the CONN_REQ to the peer has to find its EP in the EP PTR map */
+        /* Use the peer's EP PTR map key, since the EP has to send the CONN_REQ
+         * to the peer which has to find an EP created for this connection in
+         * the EP PTR map */
         ep->cm_id.ptr_map_key = addr->ptr_map_key;
         return uct_tcp_ep_connect(ep);
     }

